src/dlio/odom.cc

```diff
@@ -447,18 +449,78 @@ void dlio::OdomNode::publishToROS(pcl::PointCloud<PointType>::ConstPtr published
   br.sendTransform(transformStamped);
 }
 
+// Function for correcting the pointcloud tilestamp to more accurate represent its pose
+ros::Time dlio::OdomNode::correctTimestamp(ros::Time original_stamp) {
+  
+  // Pose buffer is empty, return previous stamp
+  if (this->pose_buffer.empty()) {
+    printf("Correct timestamp failed 1\n");
+    return original_stamp;
+  }
+
+  // Required variables
+  auto curr_pose = this->pose_buffer.begin(); // Current pose
+  auto prev_pose = this->pose_buffer.begin(); // Previous pose
+  auto best_prev_pose = this->pose_buffer.begin(); // Best previous pose
+  auto best_curr_pose = this->pose_buffer.begin(); // Best current pose
+  double best_cost = -1;
+
+  // Loop through buffer
+  while (curr_pose->second.toSec() > original_stamp.toSec()) {
+    
+    // End of buffer reached but scan time not reached yet, return previous stamp (end-1 is last valid entry)
+    if (curr_pose == this->pose_buffer.end()-1) {
+      printf("Correct timestamp failed 2\n");
+      return original_stamp;
+    }
+
+    // Update iterators
+    prev_pose = curr_pose;
+    curr_pose++;
+    
+    // Find the sum of distances between lidarPose and 2 consecutive poses
+    double cost = (prev_pose->first - this->lidarPose.p).norm() + (curr_pose->first - this->lidarPose.p).norm();
+    
+    // Save cost and pose if better than previous best
+    if (best_cost == -1 || best_cost > cost) {
+      best_cost = cost;
+      best_prev_pose = prev_pose;
+      best_curr_pose = curr_pose;
+    }
+  }
+  // We now know the best previous and next pose to the lidar pose and can interpolate the timestamp
+  double dt = best_prev_pose->second.toSec() - best_curr_pose->second.toSec();
+  float prev_weight = (best_prev_pose->first - this->lidarPose.p).norm() / best_cost;
+  double newTime = dt * prev_weight + best_curr_pose->second.toSec();
+  //printf("%lf\n%lf\n%f\n%f\n%f\n", best_prev_pose->second.toSec(), best_curr_pose->second.toSec(), prev_weight, dt, newTime);
+  return ros::Time().fromSec(newTime);
+}
+
 void dlio::OdomNode::publishCloud(pcl::PointCloud<PointType>::ConstPtr published_cloud, Eigen::Matrix4f T_cloud) {
 
   if (this->wait_until_move_) {
     if (this->length_traversed < 0.1) { return; }
   }
-
+  this->tempPose2.p = this->lidarPose.p;
+  this->tempPose2.q = this->lidarPose.q;
+  Eigen::Matrix4f transform_matrix = Eigen::Matrix4f::Identity();
+  transform_matrix.block(0,0,3,3) = (this->tempPose2.q.normalized() * this->tempPose1.q.normalized().inverse()).normalized().toRotationMatrix();
+  transform_matrix(0,3) = this->tempPose2.p[0] - this->tempPose1.p[0];
+  transform_matrix(1,3) = this->lidarPose.p[1] - this->tempPose1.p[1];
+  transform_matrix(2,3) = this->lidarPose.p[2] - this->tempPose1.p[2];
```

Comment:
I assume this is supposed to be all `tempPose2`? But it is unclear to me what you're doing here. `tempPose1` seems to be set at [line 615-616](https://github.com/Formula-Electric-Belgium/direct_lidar_inertial_odometry/blob/04da041cddc6894b16d42c35a1933e4fa11029cd/src/dlio/odom.cc#L615-L616), but the value is also `lidarPose`? So nothing happens because of this transform?

At the moment we have the following issue:
![image](https://github.com/Formula-Electric-Belgium/direct_lidar_inertial_odometry/assets/20578313/4cd40f39-7d0b-4bc4-9e45-111750c90a7b)
Black = `this->state`, green = `this->lidarPose`
The state is interpolated by calculations and new data from the IMU, but the eventual path will always be a bit off from the POV of the actual lidar position (this is to smooth out the path). So the transform we want to apply is one from the `lidarPose` to the `state`, so that the scan is seen in the pov of the `lidarPose` from the `state` position.

The calculation if this transform can also be done around [line 862](https://github.com/Formula-Electric-Belgium/direct_lidar_inertial_odometry/blob/04da041cddc6894b16d42c35a1933e4fa11029cd/src/dlio/odom.cc#L862C5-L862C5). This removes the need for `tempPose1` and `tempPose2` and the transform can be directly applied to the transform passed to the `publishToROS` function on [line 866](https://github.com/Formula-Electric-Belgium/direct_lidar_inertial_odometry/blob/04da041cddc6894b16d42c35a1933e4fa11029cd/src/dlio/odom.cc#L866).